algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-                    .sum::<F>();
-                (circuit_id, combined_x_at_beta)
+                    .map(|x| x.iter().zip_eq(&lag_at_beta).map(|(x, l)| *x * l).sum::<F>())
+                    .collect_vec();
+                (circuit_id, x_at_beta)
             })
             .collect::<BTreeMap<_, _>>();
 
-        // We're now going to calculate the lincheck_sumcheck
-        // Note that we precalculated a number of terms already to prevent duplicate calculations:
+        let g_1_at_beta = evals.get_lc_eval(&g_1, beta)?;
+
+        // We're now going to calculate the lineval_sumcheck
         #[rustfmt::skip]
-        let lincheck_sumcheck = {
-            let mut lincheck_sumcheck = LinearCombination::empty("lincheck_sumcheck");
+        let lineval_sumcheck = {
+            let mut lineval_sumcheck = LinearCombination::empty("lineval_sumcheck");
             if MM::ZK {
-                lincheck_sumcheck.add(F::one(), "mask_poly");
+                lineval_sumcheck.add(F::one(), "mask_poly");
             }
-            for (id, c) in batch_combiners.iter() {
-                let mut circuit_term = LinearCombination::empty(format!("lincheck_sumcheck term {id}"));
+            for (i, (id, c)) in batch_combiners.iter().enumerate() {
+                let mut circuit_term = LinearCombination::empty(format!("lineval_sumcheck term {id}"));
+                let fourth_sums_i = &sums_fourth_msg[i];
+                let circuit_state = &state.circuit_specific_states[id];
+
                 for (j, instance_combiner) in c.instance_combiners.iter().enumerate() {
-                    let z_a_j = witness_label(*id, "z_a", j);
                     let w_j = witness_label(*id, "w", j);
-                    circuit_term
-                        .add(r_alpha_at_beta_s[id] * instance_combiner * (eta_a + eta_c * z_b_s_at_beta[id][j]), z_a_j)
-                        .add(-t_at_beta_s[id] * v_X_at_beta[id] * instance_combiner, w_j);
+                    let mut lineval = LinearCombination::empty(format!("lineval term {j}"));
+                    let sum_a_fourth = fourth_sums_i.sum_a * circuit_state.non_zero_a_domain.size_as_field_element;
+                    let sum_b_fourth = fourth_sums_i.sum_b * circuit_state.non_zero_b_domain.size_as_field_element;
+                    let sum_c_fourth = fourth_sums_i.sum_c * circuit_state.non_zero_c_domain.size_as_field_element;
+
+                    lineval.add(sum_a_fourth * x_at_betas[id][j], LCTerm::One);
+                    lineval.add(sum_a_fourth * v_X_at_beta[id], w_j.clone());
+
+                    lineval.add(sum_b_fourth * eta_b * x_at_betas[id][j], LCTerm::One);
+                    lineval.add(sum_b_fourth * eta_b * v_X_at_beta[id], w_j.clone());
+
+                    lineval.add(sum_c_fourth * eta_c * x_at_betas[id][j], LCTerm::One);
+                    lineval.add(sum_c_fourth * eta_c * v_X_at_beta[id], w_j.clone());
```

Comment:
```suggestion
                    lineval.add(sum_c_fourth * eta_c * v_X_at_beta[id], w_j);
```
this last `clone` should be avoidable

Reply:
d1347c462